### PR TITLE
Fix websocket binary message support

### PR DESCRIPTION
**Pull request by abdallag**
Add support for base64 codec of binary websocket messages similar to proxy.

### CODE REVIEW

File: agent/websockets/shim.go [hunk 1]
```diff
@@ -272,12 +273,14 @@ func ShimBody(shimPath string) (func(resp *http.Response) error, error) {
...
 	ID      string      `json:"id,omitempty"`
 	Message interface{} `json:"msg,omitempty"`
+	Version int         `json:"v,omitempty"`
 }
 
 func createShimChannel(ctx context.Context, host, shimPath string, rewriteHost bool, openWebsocketWrapper func(http.Handler) http.Handler) http.Handler {
 	var connections sync.Map
 	var sessionCount uint64
 	mux := http.NewServeMux()
+	var protocolVersion int
```
Comment (ojarjur) [inline agent/websockets/shim.go:283]: You have this as a single, mutable variable in the proxy agent that gets updated every time a new client opens up a websocket-shim connection.

However, I don't think that will work, as different clients might be speaking different versions of the protocol.

Instead, I would recommend adding this as a field of the `Connection` struct. That way the protocol version will be determined separately for each new websocket-shim connection based on the `open` request sent by the client, and that version will remain consistent for as long as the connection is open.

**Review by ojarjur — COMMENTED** [review]
I would like to see an extension to the client-side JS shim code to use the new protocol version.

File: agent/websockets/shim.go (hunk 1 above)
Comment (abdallag) [reply agent/websockets/shim.go:299]: Good catch! Done.

**Comment by abdallag** [comment]
Added fix in websocket shim java script.

File: agent/websockets/connection.go [hunk 2]
```diff
@@ -175,7 +182,7 @@ func (conn *Connection) Close() {
 // SendClientMessage sends the given message to the websocket server.
 //
 // The returned error value is non-nill if the connection has been closed.
-func (conn *Connection) SendClientMessage(msg interface{}) error {
+func (conn *Connection) SendClientMessage(msg interface{}, ver int) error {
```
Comment (ojarjur) [inline agent/websockets/connection.go:185]: This `ver` argument isn't used (and does not seem necessary)

File: agent/websockets/connection.go [hunk 3]
```diff
@@ -208,20 +227,20 @@ func (conn *Connection) SendClientMessage(msg interface{}) error {
 //
 // The returned []string value is nil if the error is non-nil, or if the method
 // times out while waiting for a server message.
-func (conn *Connection) ReadServerMessages() ([]interface{}, error) {
+func (conn *Connection) ReadServerMessages(protocolVersion int) ([]interface{}, error) {
```
Comment (ojarjur) [inline agent/websockets/connection.go:230]: This new `protocolVersion` method argument does not seem necessary

File: agent/websockets/shim.go [hunk 4]
```diff
@@ -272,12 +289,14 @@ func ShimBody(shimPath string) (func(resp *http.Response) error, error) {
...
```
Comment (ojarjur) [inline agent/websockets/shim.go:299]: This variable is dead code now and should be removed.

File: agent/websockets/shim.go [hunk 5]
```diff
@@ -132,26 +140,30 @@ const (
...
-        }
-        return true;
+		if (Array.isArray(msg)) {
+		  if (msg.length == 1) {
+	  		if (typeof msg[0] == 'string') {
+  			  return false;
+			}
+		  }
+		}
+		return true;
```
Comment (ojarjur) [inline agent/websockets/shim.go:150]: Somehow the whitespace got mangled here (maybe an issue with spaces being replaced with tabs?)

File: agent/websockets/shim.go [hunk 6]
```diff
@@ -132,26 +140,30 @@ const (
...
+                msgs[i].msg = [btoa(reader.result)];
+            } else {
+                msgs[i].msg = [reader.result];
+            }
+            self.convertMessagesAndPush(msgs);
             });
-            reader.readAsText(blob);
-            return;
+			reader.readAsText(blob);
+			return;
```
Comment (ojarjur) [inline agent/websockets/shim.go:166]: Similarly, here the whitespace got mangled.

File: agent/websockets/connection.go (hunk 2 above)
Comment (abdallag) [reply agent/websockets/connection.go]: There were merge issues in the last commit. Now fixed in the latest.

File: agent/websockets/connection.go (hunk 3 above)
Comment (abdallag) [reply agent/websockets/connection.go]: Merge issues as mentioned above.

File: agent/websockets/shim.go (hunk 5 above)
Comment (abdallag) [reply agent/websockets/shim.go]: Fixed

File: agent/websockets/shim.go (hunk 6 above)
Comment (abdallag) [reply agent/websockets/shim.go]: Fixed

File: agent/websockets/shim.go (hunk 4 above)
Comment (abdallag) [reply agent/websockets/shim.go]: Merge issues as mentioned above.

**Review by ojarjur — APPROVED** [review]

**Pull request merged**